client/src/main/java/io/pravega/client/stream/impl/ControllerResolverFactory.java

```diff
@@ -161,26 +158,60 @@ public void shutdown() {
         private void getControllers() {
+            log.debug("Attempting to refresh the controller server endpoints");
             try {
-                final ServerResponse controllerServerList =
-                        this.client.getControllerServerList(ServerRequest.getDefaultInstance());
-                final ResolvedServerInfoGroup serverInfoGroup = ResolvedServerInfoGroup.builder()
-                        .addAll(controllerServerList.getNodeURIList()
-                                .stream()
-                                .map(node ->
-                                        new ResolvedServerInfo(
-                                                new InetSocketAddress(node.getEndpoint(), node.getPort())))
-                                .collect(Collectors.toList()))
-                        .build();
+                final ResolvedServerInfoGroup serverInfoGroup;
+                if (this.enableDiscovery) {
+                    final ServerResponse controllerServerList =
+                            this.client.getControllerServerList(ServerRequest.getDefaultInstance());
+                    serverInfoGroup = ResolvedServerInfoGroup.builder()
+                            .addAll(controllerServerList.getNodeURIList()
+                                    .stream()
+                                    .map(node ->
+                                            new ResolvedServerInfo(
+                                                    new InetSocketAddress(node.getEndpoint(), node.getPort())))
+                                    .collect(Collectors.toList()))
+                            .build();
+                } else {
+                    // Resolve the bootstrapped server list to get the set of controllers.
+                    final ArrayList<InetSocketAddress> resolvedAddresses = new ArrayList<>();
+                    this.bootstrapServers.forEach(address -> {
+                        try {
+                            resolvedAddresses.add(new InetSocketAddress(InetAddress.getByName(address.getHostString()),
+                                    address.getPort()));
+                        } catch (UnknownHostException e) {
+                            log.warn("Couldn't resolve controller address: {}, skipping this controller",
+                                    address.getHostString());
+                        }
+                    });
+                    serverInfoGroup = ResolvedServerInfoGroup.builder().addAll(resolvedAddresses.stream()
+                            .map(ResolvedServerInfo::new)
+                            .collect(Collectors.toList()))
+                            .build();
+                }
 
                 // Update gRPC load balancer with the new set of server addresses.
                 log.info("Updating client with controllers: {}", serverInfoGroup);
                 this.resolverUpdater.onUpdate(Collections.singletonList(serverInfoGroup), Attributes.EMPTY);
```

Comment:
ResolverUpdator is accessed from multiple threads and is not guarded by a lock here, so this might result in calling `onUpdate` on an uninitialized object. Either make resolverUprator final, or make it guarded with a lock everywhere and annotate it accordingly.